### ISSUE TITLE
gocd-server VOLUME /var/go makes no sense

### ISSUE THREAD

**Issue by holybit** [gocd/gocd-docker#21]
**TL;DR**
Remove `/etc/go/` and `/var/go` from gocd-server [Dockerfile.gocd-server](https://github.com/gocd/gocd-docker/blob/master/Dockerfile.gocd-server#L18) as in it's current state it doesn't allow use of `gocd/gocd-server` as a base image which can be configured (e.g., cruise-config.xml, ssh keys for go user, etc.)

I imagine the sorts of files one would put into `/var/go` would be configuration related. For example, I am using the[ GoCD SCM plugin for GitHub](https://github.com/ashwanthkumar/gocd-build-github-pull-requests). For GoCD to pull code from GitHub you have to have a private key in the gocd-server Docker container. As the server is running as the unix go user the logical location is `/var/go/.ssh/id_rsa` with the public half of the key pair in GitHub. The typical ways you would get the private key into the gocd-server docker image or container are as follows:
1. `docker run -d -v id_rsa:/var/go/.ssh/id_rsa gocd/gocd-server`
2. Build a custom image `FROM gocd/gocd-server` with the private key copied into position

Option one will not work as the owner of /var/go/.ssh/id_rsa will be root. Option two gives us the flexibility of changing the id_rsa owner and permission (i.e, go user owned and 600) so SSH works correctly. But this line from [Dockerfile.gocd-server](https://github.com/gocd/gocd-docker/blob/master/Dockerfile.gocd-server#L18) is causing issues:

``` Dockerfile
VOLUME ["/var/lib/go-server", "/var/log/go-server", "/etc/go", "/go-addons", "/var/go"]
```

Docker data volumes are intended for data not configs which often need to be dynamically modified. Here is an example of what happens if you try to use `gocd/gocd-server` as a base image and modify `/var/go`

``` Dockerfile
FROM gocd/gocd-server:16.1.0

RUN mkdir /var/go/.ssh \
    && ls -al
```

``` bash
$ docker version
Client:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   cedd534-dirty
 Built:        Tue Feb  2 13:28:10 UTC 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   cedd534-dirty
 Built:        Tue Feb  2 13:28:10 UTC 2016
 OS/Arch:      linux/amd64

$ docker build -t foo .
Sending build context to Docker daemon 10.24 kB
Step 1 : FROM gocd/gocd-server:16.1.0
 ---> f9088466c94b
Step 2 : RUN mkdir /var/go/.ssh     && ls -al /var/go/
 ---> Running in 07ccb25529ea
total 48
drwxr-xr-x.  3 go   go   4096 Feb 13 00:07 .
drwxr-xr-x. 12 root root 4096 Feb 12 22:52 ..
-rw-r--r--.  1 go   go    220 Apr  9  2014 .bash_logout
-rw-r--r--.  1 go   go   3637 Apr  9  2014 .bashrc
-rw-r--r--.  1 go   go    675 Apr  9  2014 .profile
drwxr-xr-x.  2 root root 4096 Feb 13 00:07 .ssh
 ---> 829bb624192e
Removing intermediate container 07ccb25529ea
Successfully built 829bb624192e

$ docker run -it --rm foo bash
root@70e85e684ff2:/tmp# ls -al /var/go/
total 40
drwxr-xr-x.  2 go   go   4096 Feb 13 00:07 .
drwxr-xr-x. 12 root root 4096 Feb 12 22:52 ..
-rw-r--r--.  1 go   go    220 Apr  9  2014 .bash_logout
-rw-r--r--.  1 go   go   3637 Apr  9  2014 .bashrc
-rw-r--r--.  1 go   go    675 Apr  9  2014 .profile
```

**Comment by holybit** [comment]
Docker issue [#3639](https://github.com/docker/docker/issues/3639) pinpoints this problem. From the sound of things in the issue `VOLUME` is behaving correctly.

To repeat, directories used for configs such as `/etc/go` and `/var/go` should not be declared in a base image with `VOLUME` as they can not be modified by another `Dockerfile` that inherits from the base image as instructions like `RUN` (e.g., `RUN mkdir /var/go/.ssh`) will have no affect.

**Comment by holybit** [comment]
PR #24 submitted

**Comment by holybit** [comment]
Any update on this issue from the Thought Works folks?

**Comment by an-tex** [comment]
+1

**Comment by boroborome** [comment]
+1

**Comment by pedrocarrico** [comment]
I'm also having an issue that "/var/go" is exposed as a VOLUME and I cannot easily copy my ssh keys or change any configuration in there.

👍

**Comment by arvindsv** [comment]
Closing. This has changed significantly over time. Also, there is no VOLUME definition now.
